p2p/events.py

```diff
@@ -8,6 +9,36 @@
 )
 
 
+class BaseDiscoveryServiceResponse(BaseEvent):
+
+    def __init__(self, error: Exception) -> None:
+        self.error = error
+
+
+class PeerCandidatesResponse(BaseDiscoveryServiceResponse):
+
+    def __init__(self, candidates: Tuple[str, ...], error: Exception=None) -> None:
```

Comment:
*might* be nice to use a custom `TypeVar` to represent the `str` representation of a `Node`

Reply:
I guess you mean a `NewType` . Yes we *could* do that given how "short-lived" and "fenced" these `str` are I'm not sure if the extra `NewType` is worth it. It's really just a very small path that these travel as strings and really only because of the serialization issue.